### PR TITLE
cam6_3_082: Abstract aerosol interfaces in ice nucleate

### LINKED ISSUE #648: Refactor ice nucleation code to use abstract aerosol interfaces
### What is the feature/what would you like to discuss?

Refactor `nucleate_ice_cam.F90` to use abstract aerosol interfaces. This is the 2nd step (of many) to address issue https://github.com/ESCOMP/CAM/issues/495 (integrating CARMA trop_strat model into CAM physics).

### Is there anyone in particular you want to be part of this conversation?

_No response_

### Will this change (regression test) answers?

I Don't Know

### Will you be implementing this enhancement yourself?

Yes

### PR DESCRIPTION
**Pull request by fvitt**
closes #648

### CODE REVIEW

**Review by gold2718 — COMMENTED** [review]
I'm trying to understand this and have a few questions.

File: src/chemistry/aerosol/aerosol_state_mod.F90 [hunk 1]
```diff
@@ -205,4 +270,121 @@ subroutine loadaer( self, aero_props, istart, istop, k,  m, cs, phase, &
...
+
+    wght(:,:) = 0._r8
+    totalmass(:,:) = 0._r8
+    mass(:,:)   = 0._r8
+
+    if (species_type=='sulfate_strat') then
+       sptype = 'sulfate'
+    else
+       sptype = species_type
+    end if
```
[the remark this answers is not in the document]
Comment (fvitt) [reply src/chemistry/aerosol/aerosol_state_mod.F90:310]: Sulfate is the only  "_strat" type for now.  I don't know how I feel about having a function that truncates the type string.   If there were more "_strat" types I could see your point.

File: src/physics/cam/microp_aero.F90 [hunk 2]
```diff
@@ -340,7 +340,11 @@ subroutine microp_aero_init(phys_state,pbuf2d)
       call add_default ('WSUB     ', 1, ' ')
    end if
 
-   call nucleate_ice_cam_init(mincld, bulk_scale, pbuf2d)
+   if (associated(aero_props_obj)) then
+      call nucleate_ice_cam_init(mincld, bulk_scale, pbuf2d, aero_props=aero_props_obj)
+   else
+      call nucleate_ice_cam_init(mincld, bulk_scale, pbuf2d)
+   end if
```
[the remark this answers is not in the document]
Comment (fvitt) [reply src/physics/cam/microp_aero.F90]: These were moved as suggested.

File: src/chemistry/aerosol/aerosol_state_mod.F90 [hunk 3]
```diff
@@ -97,6 +131,36 @@ subroutine aero_get_transported( self, transported_array )
...
+       integer, intent(in) :: ncol                ! number of columns
+       integer, intent(in) :: nlev                ! number of vertical levels
+       character(len=*), intent(in) :: species_type  ! species type
+       logical, intent(in) :: use_preexisting_ice ! pre-existing ice flag
+       real(r8), intent(out) :: wght(:,:)
+
+     end subroutine aero_icenuc_size_wght1
+
+     !------------------------------------------------------------------------------
+     ! return aerosol bin size weights for a given bin, column and verical layer
```
[the remark this answers is not in the document]
Comment (fvitt) [reply src/chemistry/aerosol/aerosol_state_mod.F90]: fixed

File: src/chemistry/aerosol/modal_aerosol_state_mod.F90 [hunk 4]
```diff
@@ -101,6 +132,16 @@ subroutine get_ambient_mmr(self, species_ndx, bin_ndx, mmr)
...
   end subroutine get_ambient_mmr
 
+  !------------------------------------------------------------------------
+  ! returns total ambient aerosol mass mixing ratio for a given bin index
+  !------------------------------------------------------------------------
+  subroutine get_ambient_total_mmr(self, bin_ndx, mmr)
+    class(modal_aerosol_state), intent(in) :: self
+    integer, intent(in) :: bin_ndx      ! bin index
+    real(r8), pointer :: mmr(:,:)       ! mass mixing ratios
+
```
[the remark this answers is not in the document]
Comment (fvitt) [reply src/chemistry/aerosol/modal_aerosol_state_mod.F90]: This is no longer relevant.  I have implemented `update_bin` method and removed these incomplete methods which where used only for the CARMA aerosols.

File: src/chemistry/aerosol/modal_aerosol_state_mod.F90 [hunk 5]
```diff
@@ -113,6 +154,16 @@ subroutine get_cldbrne_mmr(self, species_ndx, bin_ndx, mmr)
...
   end subroutine get_cldbrne_mmr
 
+  !------------------------------------------------------------------------
+  ! returns total cloud-borne aerosol mass mixing ratio for a given bin index
+  !------------------------------------------------------------------------
+  subroutine get_cldbrne_total_mmr(self, bin_ndx, mmr)
+    class(modal_aerosol_state), intent(in) :: self
+    integer, intent(in) :: bin_ndx      ! bin index
+    real(r8), pointer :: mmr(:,:)       ! mass mixing ratios
+
```
[the remark this answers is not in the document]
Comment (fvitt) [reply src/chemistry/aerosol/modal_aerosol_state_mod.F90]: No longer relevant.

File: src/chemistry/aerosol/aerosol_state_mod.F90 [hunk 6]
```diff
@@ -41,6 +50,21 @@ module aerosol_state_mod
 
   abstract interface
 
+     !------------------------------------------------------------------------
+     ! Total aerosol mass mixing ratio for a bin in a given grid box location (column and layer)
+     !------------------------------------------------------------------------
+     function aero_get_amb_total_bin_mmr(self, aero_props, bin_ndx, col_ndx, lyr_ndx) result(mmr_tot)
+       import
```
[the remark this answers is not in the document]
Comment (fvitt) [reply src/chemistry/aerosol/aerosol_state_mod.F90]: fixed

**Review by nusbaume — CHANGES_REQUESTED** [review]
Looks good!  I mostly just have some typo and parameter suggestions (and wanted to get them in before I go MIA).

File: src/physics/cam/nucleate_ice_cam.F90 [hunk 7]
```diff
@@ -655,104 +586,42 @@ subroutine nucleate_ice_cam_calc( &
...
-                        * (0.5_r8 - 0.5_r8*erf(log(0.1e-6_r8/dgnum(i,k,mode_aitken_idx))/  &
-                        (2._r8**0.5_r8*log(sigmag_aitken)))))
-                  end if
-               else
-                  ! all so4 from aitken
-                  so4_num  = num_aitken(i,k)*rho(i,k)*1.0e-6_r8
-               end if
+   else
+      ! for bulk model
+      dust_num_col(:ncol,:) = naer2(:ncol,:,idxdst1)/25._r8 *1.0e-6_r8 &
```
[the remark this answers is not in the document]
Comment (fvitt) [reply src/physics/cam/nucleate_ice_cam.F90]: Added `per_cm3` parameter.  Does this make sense?

File: src/chemistry/aerosol/aerosol_properties_mod.F90 [hunk 8]
```diff
@@ -77,6 +84,30 @@ subroutine aero_props_get(self, bin_ndx, species_ndx, density,hygro)
        real(r8), optional, intent(out) :: hygro   ! hygroscopicity
      end subroutine aero_props_get
 
+     !------------------------------------------------------------------------
+     ! returns species name
+     !------------------------------------------------------------------------
+     subroutine aero_species_name(self, bin_ndx, species_ndx, specname)
```
[the remark this answers is not in the document]
Comment (fvitt) [reply src/chemistry/aerosol/aerosol_properties_mod.F90]: This subroutine was not used.  I removed it.

File: src/chemistry/aerosol/aerosol_properties_mod.F90 [hunk 9]
```diff
@@ -114,6 +145,29 @@ subroutine aero_mmr_names(self, bin_ndx, species_ndx, name_a, name_c)
        character(len=32), intent(out) :: name_c ! constituent name of cloud-borne aerosol MMR
      end subroutine aero_mmr_names
 
+     !------------------------------------------------------------------------
+     ! returns constituent name of ambient aersol number mixing ratios
```
[the remark this answers is not in the document]
Comment (fvitt) [reply src/chemistry/aerosol/aerosol_properties_mod.F90]: fixed

File: src/chemistry/aerosol/aerosol_properties_mod.F90 [hunk 10]
```diff
@@ -114,6 +145,29 @@ subroutine aero_mmr_names(self, bin_ndx, species_ndx, name_a, name_c)
...
+     subroutine aero_amb_num_name(self, bin_ndx, name)
+       import :: aerosol_properties
+       class(aerosol_properties), intent(in) :: self
+       integer, intent(in) :: bin_ndx          ! bin number
+       character(len=32), intent(out) :: name  ! constituent name of ambient aerosol number dens
+
+     end subroutine aero_amb_num_name
+
+     !------------------------------------------------------------------------
+     ! returns constituent name of ambient aersol mass mixing ratios
```
[the remark this answers is not in the document]
Comment (fvitt) [reply src/chemistry/aerosol/aerosol_properties_mod.F90]: fixed

File: src/chemistry/aerosol/aerosol_state_mod.F90 [hunk 11]
```diff
@@ -26,12 +26,20 @@ module aerosol_state_mod
...
+     procedure(aero_get_amb_total_bin_mmr), deferred :: ambient_total_bin_mmr
      procedure(aero_get_state_mmr), deferred :: get_ambient_mmr
      procedure(aero_get_state_mmr), deferred :: get_cldbrne_mmr
      procedure(aero_get_state_num), deferred :: get_ambient_num
      procedure(aero_get_state_num), deferred :: get_cldbrne_num
      procedure(aero_get_states), deferred :: get_states
+     procedure(aero_update_bin), deferred :: update_bin
      procedure :: loadaer
+     procedure(aero_icenuc_size_wght1), deferred :: icenuc_size_wght1
+     procedure(aero_icenuc_size_wght2), deferred :: icenuc_size_wght2
```
[the remark this answers is not in the document]
Comment (fvitt) [reply src/chemistry/aerosol/aerosol_state_mod.F90]: Used your suggestions on the names.

File: src/physics/cam/nucleate_ice_cam.F90 [hunk 12]
```diff
@@ -170,29 +154,83 @@ end subroutine nucleate_ice_cam_register
...
+   if (prog_modal_aero.and.use_preexisting_ice) then
+
+      if (.not. present(aero_props)) then
+         call endrun(routine//' :  aero_props must be present')
+      end if
+
+      ! constituent tendencies are calculated only if use_preexisting_ice is TRUE
+      ! set lq for constituent tendencies --
+
+      allocate(aer_cnst_idx(aero_props%nbins(),0:maxval(aero_props%nspecies())))
```
[the remark this answers is not in the document]
Comment (fvitt) [reply src/physics/cam/nucleate_ice_cam.F90]: Done

File: src/physics/cam/nucleate_ice_cam.F90 [hunk 13]
```diff
@@ -170,29 +154,83 @@ end subroutine nucleate_ice_cam_register
...
+            call cnst_get_ind(tmpname, idxtmp, abort=.false.)
+            aer_cnst_idx(m,0) = idxtmp
+            if (idxtmp>0) then
+               lq(idxtmp) = .true.
+            end if
+
+            ! iterate over the species within the bin
+            do l = 1, aero_props%nspecies(m)
+               if (aero_props%icenuc_updates_mmr(m,l)) then
+                  ! this aerosol constituent will need to updated
```
[the remark this answers is not in the document]
Comment (fvitt) [reply src/physics/cam/nucleate_ice_cam.F90]: fixed

File: src/physics/cam/nucleate_ice_cam.F90 [hunk 14]
```diff
@@ -783,17 +652,85 @@ subroutine nucleate_ice_cam_calc( &
...
+                        ! iterate over the species within the bin
+                        do l = 1, aero_props%nspecies(m)
+                           if (aero_props%icenuc_updates_mmr(m,l)) then
+
+                              call aero_props%species_type(m, l, spectype)
+                              call aero_state%icenuc_size_wght( m, i,k, spectype, use_preexisting_ice, wght)
+
+                              if (wght>0._r8) then
+
+                                 ! this aerosol constituent will need to updated
```
[the remark this answers is not in the document]
Comment (fvitt) [reply src/physics/cam/nucleate_ice_cam.F90]: Corrected the grammar here

File: src/chemistry/aerosol/aerosol_state_mod.F90 [hunk 15]
```diff
@@ -205,4 +276,121 @@ subroutine loadaer( self, aero_props, istart, istop, k,  m, cs, phase, &
...
+          call aero_props%species_type(m, l, spectype)
+
+          call self%icenuc_size_wght(m, ncol, nlev, spectype, use_preexisting_ice, size_wghts)
+
+          call self%icenuc_type_wght(m, ncol, nlev, spectype, aero_props, rho, type_wghts)
+
+          select case ( trim(spectype) )
+          case('dust')
+             dust_num_col(:ncol,:) = dust_num_col(:ncol,:) &
+                  + size_wghts(:ncol,:)*type_wghts(:ncol,:)*num_col(:ncol,:)*rho(:ncol,:)*1.0e-6_r8
```
[the remark this answers is not in the document]
Comment (nusbaume) [reply src/chemistry/aerosol/aerosol_state_mod.F90:374]: Missed question/suggestion.

File: src/chemistry/aerosol/aerosol_state_mod.F90 [hunk 16]
```diff
@@ -205,4 +276,121 @@ subroutine loadaer( self, aero_props, istart, istop, k,  m, cs, phase, &
...
+    character(len=*), intent(in) :: species_type  ! species type
+    class(aerosol_properties), intent(in) :: aero_props ! aerosol properties object
+    real(r8), intent(in) :: rho(:,:)        ! air density (kg m-3)
+    real(r8), intent(out) :: wght(:,:)            ! type weights
+
+    real(r8) :: mass(ncol,nlev)
+    real(r8) :: totalmass(ncol,nlev)
+    real(r8), pointer :: aer_bin(:,:)
+
+    character(len=32) :: spectype, sptype
```
[the remark this answers is not in the document]
Comment (nusbaume) [reply src/chemistry/aerosol/aerosol_state_mod.F90:299]: Missed question/suggestion.

File: src/chemistry/aerosol/aerosol_state_mod.F90 [hunk 17]
```diff
@@ -205,4 +276,121 @@ subroutine loadaer( self, aero_props, istart, istop, k,  m, cs, phase, &
...
+                  + size_wghts(:ncol,:)*type_wghts(:ncol,:)*num_col(:ncol,:)*rho(:ncol,:)*1.0e-6_r8
+          end select
+
+       enddo
+
+       ! stratospheric sulfates
+       call self%icenuc_size_wght(m, ncol, nlev, 'sulfate_strat', use_preexisting_ice, size_wghts)
+       call self%icenuc_type_wght(m, ncol, nlev, 'sulfate_strat', aero_props, rho, type_wghts)
+       sulf_num_tot_col(:ncol,:) = sulf_num_tot_col(:ncol,:) &
+            + size_wghts(:ncol,:)*type_wghts(:ncol,:)*num_col(:ncol,:)*rho(:ncol,:)*1.0e-6_r8
```
[the remark this answers is not in the document]
Comment (nusbaume) [reply src/chemistry/aerosol/aerosol_state_mod.F90:390]: Missed question/suggestion.

File: src/physics/cam/nucleate_ice_cam.F90 [hunk 18]
```diff
@@ -514,6 +452,24 @@ subroutine nucleate_ice_cam_calc( &
...
+   integer :: idxtmp
+   real(r8), pointer :: amb_num(:,:)
+   real(r8), pointer :: amb_mmr(:,:)
+   real(r8), pointer :: cld_num(:,:)
+   real(r8), pointer :: cld_mmr(:,:)
+
+   real(r8) :: delmmr, delmmr_sum
+   real(r8) :: delnum, delnum_sum
+
+   real(r8), parameter :: per_cm3 = 1.e-6_r8
```
Comment (nusbaume) [inline src/physics/cam/nucleate_ice_cam.F90:472]: I might just add a comment here saying something like:

`Conversion factor for cubic meters to cubic centimeters`

Or something like that, just to be really clear to future users/developers.

**Review by nusbaume — CHANGES_REQUESTED** [review]
I think some of the "hidden conversation" questions/suggestions above were missed.  I tried to mark them again but not 100% sure how that will show up on your end.  Otherwise just one extra comment addition request.

File: src/chemistry/aerosol/aerosol_state_mod.F90 (hunk 15 above)
[the remark this answers is not in the document]
Comment (fvitt) [reply src/chemistry/aerosol/aerosol_state_mod.F90]: Changed as suggested

File: src/chemistry/aerosol/aerosol_state_mod.F90 (hunk 17 above)
[the remark this answers is not in the document]
Comment (fvitt) [reply src/chemistry/aerosol/aerosol_state_mod.F90]: Changed as suggested

**Review by gold2718 — APPROVED** [review]
Looks good in broad strokes (translation, I did not scrape through every last line of code).

File: src/chemistry/aerosol/aerosol_state_mod.F90 (hunk 17 above)
[the remark this answers is not in the document]
Comment (fvitt) [reply src/chemistry/aerosol/aerosol_state_mod.F90]: Turns out sulfate_strat is a special case and needs to be outside the species loop

File: src/chemistry/aerosol/aerosol_state_mod.F90 (hunk 16 above)
[the remark this answers is not in the document]
Comment (peverwhee) [reply src/chemistry/aerosol/aerosol_state_mod.F90:299]: I agree with Jesse that a parameter set to 32 would make changing these lengths in the future much simpler.

File: src/chemistry/aerosol/aerosol_state_mod.F90 [hunk 19]
```diff
@@ -205,4 +276,123 @@ subroutine loadaer( self, aero_props, istart, istop, k,  m, cs, phase, &
...
+    totalmass(:,:) = 0._r8
+    mass(:,:)   = 0._r8
+
+    if (species_type=='sulfate_strat') then
+       sptype = 'sulfate'
+    else
+       sptype = species_type
+    end if
+
+    do l = 1, aero_props%nspecies(bin_ndx)
```
Comment (peverwhee) [inline src/chemistry/aerosol/aerosol_state_mod.F90:312]: can we use a variable other than "l"? (hard to search for; looks like "1"...)

File: src/chemistry/aerosol/aerosol_state_mod.F90 [hunk 20]
```diff
@@ -205,4 +276,123 @@ subroutine loadaer( self, aero_props, istart, istop, k,  m, cs, phase, &
...
+    sulf_num_col(:,:) = 0._r8
+    soot_num_col(:,:) = 0._r8
+    sulf_num_tot_col(:,:) = 0._r8
+
+    ! collect number densities (#/cm^3) for dust, sulfate, and soot
+    do m = 1,aero_props%nbins()
+
+       call self%get_ambient_num(m, num_col)
+
+       do l = 1,aero_props%nspecies(m)
```
Comment (peverwhee) [inline src/chemistry/aerosol/aerosol_state_mod.F90:365]: same here - I'd prefer a different variable.

File: src/physics/cam/nucleate_ice_cam.F90 [hunk 21]
```diff
@@ -170,29 +154,86 @@ end subroutine nucleate_ice_cam_register
...
+            end if
+
+            call cnst_get_ind(tmpname, idxtmp, abort=.false.)
+            aer_cnst_idx(m,0) = idxtmp
+            if (idxtmp>0) then
+               lq(idxtmp) = .true.
+            end if
+
+            ! iterate over the species within the bin
+            do l = 1, aero_props%nspecies(m)
```
Comment (peverwhee) [inline src/physics/cam/nucleate_ice_cam.F90:220]: here, too? A variable other than "l"?

File: src/physics/cam/nucleate_ice_cam.F90 [hunk 22]
```diff
@@ -783,17 +657,85 @@ subroutine nucleate_ice_cam_calc( &
...
+
+                     call aero_state%get_ambient_num(m, amb_num)
+                     call aero_state%get_cldbrne_num(m, cld_num)
+
+                     if (amb_num(i,k)>0._r8) then
+                        delmmr_sum = 0._r8
+                        delnum_sum = 0._r8
+
+                        ! iterate over the species within the bin
+                        do l = 1, aero_props%nspecies(m)
```
Comment (peverwhee) [inline src/physics/cam/nucleate_ice_cam.F90:681]: and this l?

**Review by peverwhee — APPROVED** [review]
Looks good - you can take or leave my suggestions about the variable name "l"

File: src/chemistry/aerosol/aerosol_properties_mod.F90 [hunk 23]
```diff
@@ -77,6 +84,30 @@ subroutine aero_props_get(self, bin_ndx, species_ndx, density,hygro)
...
 
+     !------------------------------------------------------------------------
+     ! returns species name
+     !------------------------------------------------------------------------
+     subroutine aero_species_name(self, bin_ndx, species_ndx, specname)
+       import :: aerosol_properties
+       class(aerosol_properties), intent(in) :: self
+       integer, intent(in) :: bin_ndx           ! bin number
+       integer, intent(in) :: species_ndx       ! species number
+       character(len=32), intent(out) :: specname ! species name
```
[the remark this answers is not in the document]
Comment (fvitt) [reply src/chemistry/aerosol/aerosol_properties_mod.F90]: done

File: src/physics/cam/nucleate_ice_cam.F90 (hunk 18 above)
Comment (fvitt) [reply src/physics/cam/nucleate_ice_cam.F90]: done

File: src/chemistry/aerosol/aerosol_state_mod.F90 (hunk 19 above)
Comment (fvitt) [reply src/chemistry/aerosol/aerosol_state_mod.F90]: done

File: src/chemistry/aerosol/aerosol_state_mod.F90 (hunk 20 above)
Comment (fvitt) [reply src/chemistry/aerosol/aerosol_state_mod.F90:365]: done

File: src/physics/cam/nucleate_ice_cam.F90 (hunk 21 above)
Comment (fvitt) [reply src/physics/cam/nucleate_ice_cam.F90]: done

File: src/physics/cam/nucleate_ice_cam.F90 (hunk 22 above)
Comment (fvitt) [reply src/physics/cam/nucleate_ice_cam.F90:681]: done

**Pull request merged**